### DISCUSSION TITLE
Programmatically creating a codespaces vm for an user.

### DISCUSSION THREAD

**Discussion by artificial-cassiano** [community/community#9020] [Codespaces]
Hi, 
Is there current support for the following scenario: 
* create, programmatically, a CodeSpace accessible by one(or many) of my collaborators on a repository. 
* retrieve the URL for the, generated, CodeSpace

How would that be possible? I looked around the `gh` cli and the graphQL API but neither seem to be able to do that.

**Comment by asciimike** [comment]
@artificial-cassiano we don't currently have the ability for org admins to create codespaces on behalf of users in their org, though this is something we're considering (as you might have noticed we have [user level APIs](https://docs.github.com/en/rest/reference/codespaces) to control their codespaces, but not org ones).

If it's feasible for you to create an OAuth app and let users create/manage resources themselves (and potentially also let you manage resources on their behalf), that might be one way to solve this.

Org level APIs are definitely something we've considered; the create flow is an interesting one since we're not entirely certain where to put it:
 - `POST /users/:user/codespaces -d {user: <user>, repo:<org/repo>, branch: <branch>}`
 - `POST /repos/:org/:repo/codespaces -d {user: <user>, branch: <branch>}`
 - `POST /orgs/:org/codespaces -d {user: <user>, repo:<org/repo>, branch: <branch>}`

Successful creation would return a URL to the codespace in the response body.

I think that other operations (e.g. list all codespaces in my org) would definitely go at the org level (e.g. `GET /orgs/:org/codespaces`) though getting or deleting an individual codespace is an interesting; potentially it could just be `DELETE /orgs/:org/codespaces/:codespace_name` though I could see an argument to allow admins of an org to delete a codespace by using the existing delete endpoint `DELETE /users/codespaces/:codespace_name` so long as that codespace is owned by the org.

Would love thoughts on this as well as any more detail on what you're hoping to accomplish here!
